### PR TITLE
feat: dynamic type loading

### LINKED ISSUE #678: Dynamic gametest type fetching
Move gametest types out of bundle and fetch them on demand

### LINKED ISSUE #710: Update scripting types
Certain module versions have left experimental
https://learn.microsoft.com/en-us/minecraft/creator/documents/scriptversioning

### PR DESCRIPTION
**Pull request by Joelant05**
# Overview
This PR introduces the ability for type definitions to be fetched from external URLs at runtime. They are cached for faster access in the future. This will make it considerably faster to update type definitions.

closes #710
closes #678

# Data Format
A new type definition entry could look like this, with `moduleName` optionally defining the name of the modules these types should be available within:
```json
{
			"moduleName": "@minecraft/server",
			"definition": "https://cdn.jsdelivr.net/npm/@minecraft/server@1.0.0-preview.1.19.60.22/index.d.ts",
			"requires": {
				"dependencies": ["@minecraft/server"],
				"targetVersion": [">=", "1.19.60"],
				"experimentalGameplay": ["!enableGameTestFramework"]
			}
		}
```

# TODO

- [x] Type caching can be unreliable
- [x] Clean up code; remove timer

### CODE REVIEW

**Comment by solvedDev** [comment]
If I open an older script API project (from 1.17.40) and open it, this no longer pulls in the correct types (uses 1.19.40). Is this intentional?

File: src/components/Data/TypeLoader.ts [hunk 1]
```diff
@@ -41,21 +46,108 @@ export class TypeLoader {
...
+								false,
+								moduleName,
+							])
+							return
+						}
+
+						// The file couldn't be fetched from cache (because it is not in index or at the path specified)
+						// So we need to fetch it
+						const res = await fetch(typeLocation).catch(() => null)
+						// TODO: Maybe set a variable (failedToFetchAtLeastOnce) to later open an information window that tells the user that some types couldn't be fetched
```
Comment (solvedDev) [inline src/components/Data/TypeLoader.ts:109]: Fixed an error here when the user is not connected to a network. We can fix the TODO comment in a followup PR

File: src/components/Data/TypeLoader.ts [hunk 2]
```diff
@@ -73,26 +165,22 @@ export class TypeLoader {
...
 
-				if (valid)
-					return <const>[definition, await this.load(definition)]
-			})
-		)
-		const filteredLibs = <(readonly [string, string])[]>(
-			libs.filter((lib) => lib !== undefined)
+					return []
+				})
+				.filter((type) => type[0]) as [string, string?][]
```
Comment (solvedDev) [inline src/components/Data/TypeLoader.ts:180]: Nice simplification here!

**Comment by Joelant05** [comment]
> If I open an older script API project (from 1.17.40) and open it, this no longer pulls in the correct types (uses 1.19.40). Is this intentional?

Yes, in bridge's data it only goes back to 1.19.40 types now. This was because the versioning before that point of the type modules was inconsistent with Minecraft versions, e.g. they were versioned as 0.1.7 and 0.1.6 etc. so it would take a while to go back and find which types correspond to which versions.
I don't think this is much of a problem since I don't think Mojang had any backwards compatibility with those older scripting versions anyway, since they were experimental.

**Comment by solvedDev** [comment]
Alright, this is good to merge from my side then 👍

**Pull request merged**